app/controllers/app.rb

```diff
@@ -28,76 +32,94 @@ class App < Roda
         router.is do
           # GET /details?query=[query]
           router.get do
-            router.halt 400 unless router.params.include? 'query'
-
-            query = router.params['query']
-
-            jobs = JobCollector.new(App.config).jobs(query)
-
-            skills = []
-
-            jobs.each do |job|
-              skills += Skiller::Skill::SkillMapper.new(job).skills
+            # Examine the query
+            begin
+              query = router.params['query']
+              if query.empty? # TODO: Use regex to avoid "   " inputs
+                flash[:error] = 'This query is empty'
+                router.redirect('/')
+              end
+            rescue exception # router.params.exclude 'query', not sure what the error type is
+              flash[:error] = 'Query is not detected'
+              router.redirect('/')
             end
 
-            # for job in jobs
-            # skill = Skiller::SkillAnalyzer::Extractor.new(job)
-            #  skill.extract
-            #  skills.append(skill)
-            # end
-
-            ## TODO: extract `Skill` from jobs if the query has not been searched
-            # skills = []
-            # for job in jobs
-            #   skill = Skiller::SkillAnalyzer::Extractor.new(jobs[0])
-            #   skill.extract
-            #   puts skill.result
-
-            #   id = 1
-            #   name = 'aaa'
-            #   job_db_id = job.job_id
-            #   salary = job.salary
-            #   skills.append(Skiller::Entity::Skill.new(id:id, name:name, job_db_id:job_db_id, salary:salary))
-            # end
-
-            ## TODO: then use `Repository::JobsSkills.create()` to put them into database
-            # Repository::JobsSkills.create(skills)
-
-            ## TODO: use `Repository::QueriesJobs.find_skills_by_query()` if the query has been searched
-            # skills.each do |skill|
-            #  puts skill.name
-            # end
-            view 'details', locals: { query: query, jobs: jobs, skills: skills }
+            # Extract information and map to view object
+            collector = DataCollector.new(App.config, query)
+            jobs = collector.jobs
+            skills = collector.skills
+            skillset = Views::SkillJob.new(jobs, skills)
+
+            begin
+              flash[:notice] = "Your last query is '#{query}'"
+              view 'details', locals: { query: query, skilljob: skillset }
+            rescue NoMethodError => _e
+              flash[:error] = "No skills extracted from '#{query}'"
+              router.redirect('/')
+            end
           end
         end
       end
     end
 
     # request jobs using API if the query has not saved to database
-    class JobCollector
-      def initialize(config)
+    class DataCollector
```

Comment:
Please move this class to another layer, like domain or infrastructure. It does not belong to controller.

Reply:
Has moved to service objects in this week's progress.